gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt

```diff
@@ -641,10 +631,35 @@ abstract class KspTaskJvm @Inject constructor(
             clearIncCache()
         }
         args.addChangedFiles(changedFiles)
-        super.callCompilerAsync(args, sourceRoots, inputChanges, taskOutputsBackup)
+        super.callCompilerAsync(args, kotlinSources, inputChanges, taskOutputsBackup)
     }
 
-    override fun skipCondition(): Boolean = false
+    override fun skipCondition(): Boolean = sources.isEmpty && javaSources.isEmpty
+
+    override val incrementalProps: List<FileCollection>
+        get() = listOf(
+            sources,
+            javaSources,
+            commonSourceSet,
+            classpathSnapshotProperties.classpath,
+            classpathSnapshotProperties.classpathSnapshot
+        )
+
+    @get:InputFiles
+    @get:SkipWhenEmpty
+    @get:IgnoreEmptyDirectories
+    @get:PathSensitive(PathSensitivity.RELATIVE)
+    override val sources: FileCollection = super.sources.filter {
```

Comment:
Nit (i.e., not necessary to fix now or follow-up): Some parameter below is renamed to `kotlinSources` (from `sourceRoots`); also there is `javaSources` below. So... I guess this is `kotlinSources` as well? Not familiar what `super` is here, and maybe need to be renamed from there. Or... is there a convention about this?

Reply:
It is actually quite messy here.. the names are mostly from upstream kotlin gradle plugin therefore we need to follow these naming.